File: lib/lab_sound/oscillator_node.dart
import 'lab_sound.dart';

enum OscillatorType {
  OSCILLATOR_NONE,
  SINE,
  FAST_SINE,
  SQUARE,
  SAWTOOTH,
  TRIANGLE,
  CUSTOM,
}

class OscillatorNode extends AudioScheduledSourceNode {

  late AudioParam amplitude;
  late AudioParam frequency;
  late AudioParam detune;
  late AudioParam bias;

  OscillatorNode(AudioContext ctx)
      : super(ctx, LabSound().createOscillatorNode(ctx.pointer)) {
    amplitude = AudioParam(this.ctx, this.nodeId, LabSound().OscillatorNode_amplitude(this.nodeId));
    frequency = AudioParam(this.ctx, this.nodeId, LabSound().OscillatorNode_frequency(this.nodeId));
    detune = AudioParam(this.ctx, this.nodeId, LabSound().OscillatorNode_detune(this.nodeId));
    bias = AudioParam(this.ctx, this.nodeId, LabSound().OscillatorNode_bias(this.nodeId));
  }

  OscillatorType get type => OscillatorType.values[LabSound().OscillatorNode_type(nodeId)];

  setType(OscillatorType type) {
    LabSound().OscillatorNode_setType(nodeId, type.index);
  }

  @override
  dispose() {
    super.dispose();
  }
}

File: bridge/OscillatorNode.cpp
#include "./dart_api/dart_api.h"
#include "LabSound/LabSound.h"
#include "KeepNode.cpp"
#include "struct.h"
using namespace lab;

DART_EXPORT int createOscillatorNode(AudioContext* context) {
    auto node = std::make_shared<OscillatorNode>(*context);
    return keepNode(node);
}

DART_EXPORT int OscillatorNode_type(int nodeId) {
    auto node = std::static_pointer_cast<OscillatorNode>(getNode(nodeId));
    return node ? static_cast<int>(node->type()) : 0;
}

DART_EXPORT void OscillatorNode_setType(int nodeId, int type) {
    auto node = std::static_pointer_cast<OscillatorNode>(getNode(nodeId));
    if(node) node->setType(OscillatorType(type));
}

DART_EXPORT int OscillatorNode_amplitude(int nodeId) {
    auto node = std::static_pointer_cast<OscillatorNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 1, node->amplitude()) : -1;
}

DART_EXPORT int OscillatorNode_frequency(int nodeId) {
    auto node = std::static_pointer_cast<OscillatorNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 2, node->frequency()) : -1;
}

DART_EXPORT int OscillatorNode_detune(int nodeId) {
    auto node = std::static_pointer_cast<OscillatorNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 3, node->detune()) : -1;
}

DART_EXPORT int OscillatorNode_bias(int nodeId) {
    auto node = std::static_pointer_cast<OscillatorNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 4, node->bias()) : -1;
}
